### PR TITLE
Add support for mixtures

### PR DESCRIPTION
**Pull request by aisopous**
Implements constructors and marginalisation for Distributions.MixtureModel, and a UnionAll type KeyedMixtureModel

https://github.com/invenia/KeyedDistributions.jl/issues/5

### CODE REVIEW

File: src/KeyedDistributions.jl [hunk 1]
```diff
@@ -104,6 +107,15 @@ function Base.getindex(d::KeyedGenericMvTDist, i::Vector)::KeyedGenericMvTDist
     return KeyedDistribution(GenericMvTDist(d.d.df, d.d.μ[i], submat(d.d.Σ, i)), axiskeys(d)[1][i])
 end
 
+function Base.getindex(mm::KeyedMixtureModel, i::Vector)::KeyedMixtureModel
+    margcomps = map(mm.d.components) do c
+        T = typeof(c)
+        T.name.wrapper(mean(c)[i], Symmetric(cov(c)[i, i]))
```
[the remark this answers is not in the document]
Comment (oxinabox) [reply src/KeyedDistributions.jl:113]: Grabbing `name.wrapper` takes the type parameter away.
I imagine this is for something like `MvNormal{Int}` that should be a `MvNormal{Float64}` after marginalizing.
And the plain `MvNormal` constructor without type-params will handle working that out.
(or something roughly like that)

File: src/KeyedDistributions.jl [hunk 2]
```diff
@@ -9,7 +9,7 @@ using PDMatsExtras: submat
 using Random: AbstractRNG
 
 export KeyedDistribution, KeyedSampleable
-export KeyedMvNormal, KeyedGenericMvTDist, MvNormalLike, MvTLike
+export KeyedMvNormal, KeyedGenericMvTDist, MvNormalLike, MvTLike, KeyedMixtureModel
```
Comment (oxinabox) [inline src/KeyedDistributions.jl:12]:
```suggestion
export KeyedMvNormal, KeyedGenericMvTDist, MvNormalLike, MvTLike, KeyedMixtureModel, MixtureModelLike
```
Apparently 2we export thing thing `XLike`

File: src/KeyedDistributions.jl [hunk 3]
```diff
@@ -91,6 +91,9 @@ const MvNormalLike = Union{MvNormal, KeyedMvNormal}
 const KeyedGenericMvTDist = KeyedDistribution{Multivariate, Continuous, <:GenericMvTDist}
 const MvTLike = Union{GenericMvTDist, KeyedGenericMvTDist}
 
+const KeyedMixtureModel = KeyedDistribution{<:VariateForm, <:ValueSupport, <:AbstractMixtureModel}
+const KeyedMixtureLike = Union{MixtureModel, KeyedMixtureModel}
```
Comment (oxinabox) [inline src/KeyedDistributions.jl:95]: Naming pattern elsewhere seems to follow"
```suggestion
const MixtureLike = Union{MixtureModel, KeyedMixtureModel}
```

File: src/KeyedDistributions.jl [hunk 4]
```diff
@@ -218,4 +236,35 @@ function Distributions.insupport(d::KeyedDistribution{<:Univariate}, x::Real)
...
+function Distributions.MixtureModel(cs::Vector{C}, pri::CT) where {C<:KeyedDistribution,CT<:AbstractVector{<:Real}}
+    return MixtureModel(cs, Categorical(pri))
+end
+
+KeyedMixtureModel(cs::Vector{<:KeyedDistribution}, pri::Union{AbstractVector{<:Real}, Distributions.Categorical}) = MixtureModel(cs, pri)
+
+KeyedMixtureModel(mm::MixtureModel, keys::Tuple{Vararg{AbstractVector}}) = KeyedDistribution(mm, keys)
+
+# Avoid the double wrap
+KeyedDistribution(kd::KeyedDistribution, keys::Tuple{Vararg{AbstractVector{T} where T, N} where N}) = KeyedDistribution(kd.d, keys)
```
Comment (oxinabox) [inline src/KeyedDistributions.jl:259]: I don't recall how KeyedArray handle this.
But is this consistent with that?
Or should this be `rekey` ?

File: src/KeyedDistributions.jl [hunk 5]
```diff
@@ -218,4 +236,35 @@ function Distributions.insupport(d::KeyedDistribution{<:Univariate}, x::Real)
...
+
+function (mm::KeyedMixtureModel)(keys...) 
+    margcomps = map(mm.d.components) do c
+        inds = first(map(AxisKeys.findindex, keys, axiskeys(mm)))
+        T = typeof(c)
+        T.name.wrapper(mean(c)[inds], Symmetric(cov(c)[inds, inds]))
```
Comment (oxinabox) [inline src/KeyedDistributions.jl:265]: should we push this into a helper called `_marginalize` or something

Comment (aisopous) [reply src/KeyedDistributions.jl]: Fixed -- it's still annoying to have two solutions: marginalization is generated for KeyedDistributions but I chose the components not to be Keyed as it seemed redundant.

File: src/KeyedDistributions.jl (hunk 4 above)
Comment (aisopous) [reply src/KeyedDistributions.jl:269]: rekey doesn't exist any longer. 

`KeyedArray`'s seem to stackoverflow when attempting the same thing, so higher level code must be careful not to double wrap?

File: src/KeyedDistributions.jl [hunk 6]
```diff
@@ -91,6 +92,9 @@ const MvNormalLike = Union{MvNormal, KeyedMvNormal}
 const KeyedGenericMvTDist = KeyedDistribution{Multivariate, Continuous, <:GenericMvTDist}
 const MvTLike = Union{GenericMvTDist, KeyedGenericMvTDist}
 
+const KeyedMixtureModel = KeyedDistribution{<:VariateForm, <:ValueSupport, <:AbstractMixtureModel}
+const MixtureLike = Union{MixtureModel, KeyedMixtureModel}
```
Comment (oxinabox) [inline src/KeyedDistributions.jl:96]:
```suggestion
const MixtureModelLike = Union{MixtureModel, KeyedMixtureModel}
```

**Review by oxinabox — APPROVED** [review]
clean up any remaining issues and merge when happy

**Review by rofinn — APPROVED** [review]
Yep, this looks good to go.

**Pull request merged**